rudderlabs/rudder-sdk-kotlin
# feat: implement automatic Activity tracking

## Description

This PR add the ActivityTrackingPlugin which automatically sends screen events for any activity opened inside a client app.

## Type of change


## Implementation Details

- ActivityTrackingPlugin class is added which inherits Plugin.
- This plugin is added in the pluginChain in the Analytics class for Android.
- A `trackActivities` param is added in the Configuration which enables/disables activity tracking.
- The ActivityTrackingPlugin is uses `ActivityLifecycleCallbacks` onActivityStarted callback to track activities.

## Checklist


## How to test?

Set the trackActivities param to true in the configuration of the sample app. Run the app and check for screen events of the activity opened. Sample event:
```
{
    "anonymousId": "26be140a-1ae3-4012-a36f-d94db668b204",
    "channel": "mobile",
    "context": {
        "app": {
            "build": "100",
            "name": "Rudder-Android-Libs",
            "namespace": "com.rudderstack.android.sampleapp",
            "version": "0.1.0"
        },
        "library": {
            "name": "com.rudderstack.android.sdk",
            "version": "1.0.0"
        },
        "locale": "en-US",
        "network": {
            "bluetooth": true,
            "carrier": "T-Mobile",
            "cellular": true,
            "wifi": true
        },
        "os": {
            "name": "Android",
            "version": "15"
        },
        "screen": {
            "density": 420,
            "height": 2400,
            "width": 1080
        },
        "timezone": "Asia/Kolkata"
    },
    "event": "com.rudderstack.sampleapp.mainview.MainActivity",
    "integrations": {
        "All": true
    },
    "messageId": "afeab3ed-f05e-45c2-843d-db0d9da9e0da",
    "originalTimestamp": "2024-10-15T06:23:52.513Z",
    "properties": {
        "automatic": true,
        "name": "com.rudderstack.sampleapp.mainview.MainActivity"
    },
    "receivedAt": "2024-10-15T06:24:07.861Z",
    "request_ip": "49.36.239.246",
    "rudderId": "6375253e-04d9-4db0-99d3-ddd4e4dcb7e3",
    "sentAt": "2024-10-15T06:24:02.448Z",
    "type": "screen"
}
```

## Breaking Changes


## Maintainers Checklist


## Screenshots (if applicable)

<img width="200" alt="image" src="https://github.com/user-attachments/assets/a323ab00-d3cb-4d79-84b8-71ae65cbaba4">


## Additional Context
